Fix typing idempotency bug with Annotated trees

lrytz opened the pull request:
typedAnnotated transforms an Annotated tree into a Typed tree. The
original field of the result is set to the Annotated tree.

The bug was that typedAnnotated was using the untyped Annotated tree
as original, but also set its type. When re-typing later on the same
Annotated tree, the typer would consider it as alreadyTyped. This
is incorrect, the typer needs to convert Annotated trees to Typed.

Also, the Annotated tree only had its type field set, but its children
were still untyped. This crashed the compiler lateron, non-typed trees
would get out of the typing phase.

Review by @xeno-by

lrytz committed:
Fix typing idempotency bug with Annotated trees

typedAnnotated transforms an Annotated tree into a Typed tree. The
original field of the result is set to the Annotated tree.

The bug was that typedAnnotated was using the untyped Annotated tree
as original, but also set its type. When re-typing later on the same
Annotated tree, the typer would consider it as alreadyTyped. This
is incorrect, the typer needs to convert Annotated trees to Typed.

Also, the Annotated tree only had its type field set, but its children
were still untyped. This crashed the compiler lateron, non-typed trees
would get out of the typing phase.

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -4161,7 +4161,7 @@ trait Typers extends Modes with Adaptations with Tags {
               }
             else
-              tree
+              Annotated(ann, arg1)
```
What's the value of `tree` here, and how is it different from `Annotation(ann, arg1)`? Every single time I forget how annotated trees work.

lrytz replied:
`tree` refered in fact out to the parameter of `typed1`. It is the same reference as the `atd: Annotated` parameter of `typedAnnotated`. The main difference is that we don't set the `tpe` field of the untyped tree `atd`. Instead we create a new `Annotated` instance and set its type. Also the new `Annotated` instance has the typed `arg1` as body - I think it's better to put that into the `original` than the un-typed `arg`

xeno-by replied:
Oh now I understand the problem. Btw what language construct corresponds to
the code path you fixed?

On 12 Feb 2013, at 14:01, Lukas Rytz notifications@github.com wrote:

In src/compiler/scala/tools/nsc/typechecker/Typers.scala:

> @@ -4161,7 +4161,7 @@ trait Typers extends Modes with Adaptations with Tags {
>                  case _ => Annotated(ann, arg1)
>                }
>              else
> -              tree
> -              Annotated(ann, arg1)

tree refered in fact out to the parameter of typed1. It is the same
reference as the atd: Annotated parameter of typedAnnotated. The main
difference is that we don't set the tpe field of the untyped tree atd.
Instead we create a new Annotated instance and set its type. Also the new
Annotated instance has the typed arg1 as body - I think it's better to put
that into the original than the un-typed arg

—
Reply to this email directly or view it on
GitHubhttps://github.com/scala/scala/pull/2118/files#r2977547.

xeno-by replied:
Btw looks like the code can be simplified by omitting the enclosing if and
leaving just the pattern match.

On 12 Feb 2013, at 14:01, Lukas Rytz notifications@github.com wrote:

In src/compiler/scala/tools/nsc/typechecker/Typers.scala:

> @@ -4161,7 +4161,7 @@ trait Typers extends Modes with Adaptations with Tags {
>                  case _ => Annotated(ann, arg1)
>                }
>              else
> -              tree
> -              Annotated(ann, arg1)

tree refered in fact out to the parameter of typed1. It is the same
reference as the atd: Annotated parameter of typedAnnotated. The main
difference is that we don't set the tpe field of the untyped tree atd.
Instead we create a new Annotated instance and set its type. Also the new
Annotated instance has the typed arg1 as body - I think it's better to put
that into the original than the un-typed arg

—
Reply to this email directly or view it on
GitHubhttps://github.com/scala/scala/pull/2118/files#r2977547.

lrytz replied:
I think the crash cannot be triggered by simply compiling a file. You need some plugin (or macro?) that compiles compiles some tree more than once. See the test.

Agree on removing the `if`.

xeno-by commented:
Apart from the minor question, which I think would be great to have in comments, this looks good given it doesn't break reification (which was the sole reason for me introducing this code in the first place). I'd also get a review from @dragos, since I remember us having a problem there with Scala IDE.

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -4153,15 +4153,11 @@ trait Typers extends Modes with Adaptations with Tags {
-              tree
+          val original = arg1 match {
+            case tt @ TypeTree() => Annotated(ann, tt.original)
+            // this clause is needed to correctly compile stuff like "new C @D" or "@(inline @getter)"
```
I would also add a comment for your use case here

lrytz replied:
Well, In my use case I don't care what we use as `original` for the resulting `TypeTree`. The only thing I about care is that we don't set `tpe` fields of untyped trees randomly, which is a contract that the Typer has to follow everywhere. So I don't think a comment is helpful here.

adriaanm replied:
```
val originalArg = arg1 match { case tt @ TypeTree() => tt.original case t => t }
val original = Annotated(ann, originalArg)
```

lrytz commented:
hmm.. many aborts.. PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 13442072)
:cat: Roger! Rebuilding pr-rangepos-per-commit, pr-checkin-per-commit for 8206e26f. :rotating_light:

retronym commented:
It turns out that this bug was hit in SBT when compiling against Scala 2.10.1-RC1. The actual change the triggered is the regression was a fix to `TypeTree#copyAttrs` as [part of](https://github.com/scala/scala/pull/2090/files#L4R526) SI-6187.

The pattern that triggered the bug:

``` scala
object Bug {
  def parser0: Def.Initialize[State => sbt.complete.Parser[(Seq[String], Seq[String])]] = ???
  Def./*macro call*/inputTaskDyn[Any] {
    val (x, y) = parser0./*macro call*/parsed
    ???
  }
}
```

The pattern matcher expands the call to:

``` scala
val x$1 = (EXPANSION(parser0./*macro call*/parsed) @unchecked) match {
  case (_1, _2) => Tuple2(_1, _2)
}
val x = x$1._1
val y = x$1._2
```

The outer macro, `inputTaskDyn`, uses `resetLocalAttrs` on it's argument. Somehow this reset the type tree of the scrutinee to the unexpanded macro.

/cc @xeno-by @harrah 

Mark has a workaround: don't use pattern matching to deconstruct the result of the inner macro.

I confirmed that this patch also solved the issue

lrytz commented:
Cool! Does that mean we can get the patch into 2.10.1-RC2  ---  since it fixes a regression.. :innocent:

JamesIry commented:
Please check all that apply
- [x] bug in experimental feature
- [ ] bug with localized work around

If you scored 2 or higher it's probably not a blocker for 2.10.1. :-)

Or do you think the underlying bug will break more than inner macros in pattern matching?

retronym commented:
Mark's workaround will work for him, but it won't prevent users of his macros hitting the same problem.

The general pattern which is problematic:

``` scala
someMacroThatUseResetAttrs {  
  (someOtherMacroCall: @unchecked) match { case s => s }
```

`someMacroThatUseResetAttrs` encompasses most of the macros we're developing at Typesafe. The `@unchecked` part is added automatically by the compiler when translating certain pattern matches.

While we're discussing macro impacts here, you could replace `someMacroThatUseResetAttrs` with `someCompilerTransformThatUsesResetAttrs` and probably discover new bugs.

The lesson learned should be to perform more pre-RC testing with projects like SBT...

lrytz commented:
so the bug is not related to pattern matching per se, val patterns are just probably the most common way to trigger the bug. what you need is an annotated term (`term: @annot`). it happens that val patterns always generate such a tree:

```
val pat = expr
```

desugars to:

```
val x$1 = (expr: @untyped) match { case pat => ... }
```

in addition to an annotated term, you need a situation where that term is type-checked twice. i don't think this will ever happen in a normal compilation run, but it certainly happens for macros, compiler plugins, and maybe in the presentation compiler.

workarounds:
- for val patterns, don't use them
- if you have an annotated term in your source, add an explicit type: change `term: @annot` to `term: Type @annot`

the symptom is a compiler crash in a later stage, where is not predictable, so the error messages will vary. finding the cause / workaround would be non-trivial for users.

the patch is very local.

that's about all i can say about it - there are arguments for and against putting it in, of course..

JamesIry commented:
The arguments are getting persuasive, especially the one that says "presentation compiler." @adriaanm, thoughts?

Also, @retronym and @xeno-by, could you give an LGTM at least for 2.10.x? We can then decide to backport to 2.10.1 or not separately

lrytz commented:
i said "maybe presentation compiler" because i don't know - it's very well possible that we don't have a problem there. thinking about it, i don't think the pc re-compiles trees, afaik it will always re-parse the file before type checking it. also the bug would probably have been reported earlier.

retronym commented:
Within the compiler, at least `NamesDefaults`, `Typers#adapt` and `Typers#labelDef` use variants of `resetAttrs`. They could be susceptible. I want to spend a bit more time with the patch and produce a standalone test case based on Mark's problem before I give it a thumbs up; give me a day or two.

lrytz commented:
sounds like a good plan - thanks jason

adriaanm commented:
still on the fence, but if it indeeds causes major trouble in the IDE, i'll be in favor

macro issues with workarounds will have to be fixed in 2.10.2-RC1, though

retronym commented:
I will endeavour to find a regular type-checking related bug, in that case.

adriaanm commented:
that's one attack vector -- the other would be to show that the work around is not feasible (P_occurs \* Cost_workaround)

retronym committed:
Additional test case for Lukas' fix to annotated originals.

This was inspired by the regression that Mark encountered
when upgrading SBT from 2.10.0 to 2.10.1-RC1.

retronym commented:
LGTM, pending addition of this additional test:

  https://github.com/lrytz/scala/pull/1

The workaround is to ask people not to call macros in macro arguments, which isn't really acceptable IMO, as macro  calls look like regular calls. So I'd strongly vote for inclusion in 2.10.1.

lrytz committed:
Merge pull request #1 from retronym/topic/annotatedRetyping-test

Additional test case for Lukas' fix to annotated originals.
